course_contents/2_intro_to_python/notes/04. Destructuring.md

```diff
@@ -0,0 +1,49 @@
+### Unpacking
+- unpacking is used to assign values from a collection to variables in a single statement
+- syntax for unpacking:
+
+```py
+x, y, z = collection
+```
+
+- the collection can be a list, tuple, string, etc.
+
+An example of unpacking a list:
+
+```py
+example_list = [1,2,3]
+a, b, c = example_list
+print(a, b, c)
+```
+
+### The `enumerate()` method
+- the `enumerate()` method is used to retrieve an object containing both element and its index inside the collection that is being iterated over
+- syntax for the `enumerate()` method with unpacking:
+
+```py
+for index, element in enumerate(collecion):
+    block of code
+```
+
+An example of the `enumerate()` method:
+
+```py
+collection = ['a', 'b', 'c']
+for index, element in enumerate(collection):
+    print(f"Index:{index}; Element:{element}")
+```
+
+### The `zip()` function
+- the `zip()` function is used to combine two iterables into a single one
+- new iterable is created by joining i-th element of each iterable until the end of the shorter one is reached
+- the `zip()` function returns a `zip` object
```

Comment:
Why is it iseful to know that it returns a `zip` object? What is a `zip` object?

Reply:
I think that this would again add too much to the whole size of the notes without having any real use for students. They need to be able to quickly refresh their memory of how the zip function works and what they can do with it. It might even be better to remove the sentence mentioning the zip object from the text since it doesn't mean much without explaining the zip object which will then cause too much details to appear in these notes

Reply:
I agree, I don't think talking about `zip` objects here would bring much benefit to students. Removing line 38 might just be the way to go.

However we might run into some trouble because in the example below we do `list(zip(numbers, letters))`. The obvious question is: why do we need `list()`?

Reply:
Yep, that sounds good!

Reply:
Added a note for callout, resolving this conversation